text2chem/core/formula_parser.py

```diff
@@ -75,12 +75,12 @@ def __get_composition(init_formula):
 
 
 def __parse_parentheses(init_formula, init_factor, curr_dict):
-    re_in_parentheses = r"\(((?>[^\(\)]+|(?R))*)\)\s*([-*\.\da-z\+/]*)"
+    re_in_parentheses = r"\(((?>[^\(\)]+|(?R))*)\)\s*(\([-*.\da-z+/]*\)|[-*.\da-z+/]*)"
```

Comment:
What is this change? Could you please explain?

Reply:
As I understand it, regarding the function `__parse_parentheses`, its execution logic is based on Depth-First Search (DFS). As for the regular expression `re_in_parentheses`, it matches nested balanced parentheses **expressions** and their multiplication **factors**.

Let's explain this with an example: '(Cu1/3Nb2/3)(1/4-y-x)Ti3/4-xZr2*x+yO2'

Before the changes:
- For the **first subtree**, the expression matches the substring '(Cu1/3Nb2/3)', proceeds to the next recursion, but the expression does not match. Then, the function `__get_sym_dict` parses this substring and the multiplication factor= ‘1’, resulting in `curr_dict -> {'Cu': '0.333', 'Nb': '0.667'}`.
- For the **second subtree**, it matches the substring '(1/4-y-x)', proceeds to the next recursion, but the expression does not match. Then, `__get_sym_dict` parses. `curr_dict -> {'Cu': '0.333', 'Nb': '0.667'}`, and this is where **the error occurs**. Because '(1/4-y-x)' should be considered as a factor of '(Cu1/3Nb2/3)' and should belong to a single node.

After the changes:
- For the **first node**, it matches '(Cu1/3Nb2/3)(1/4-y-x)' -> factor='1/4-y-x', proceeds to the next recursion, the expression does not match, and then `__get_sym_dict` parses formula='Cu1/3Nb2/3',  factor='1/4-y-x'. `curr_dict -> {'Cu': '0.0833-0.333*x-0.333*y', 'Nb': '0.167-0.667*x-0.667*y'}`.

Reply:
Thanks! I understood now. 👍